### PR TITLE
Implement xod-dev/ws2812

### LINKED ISSUE #1921: Implement xod-dev/ws2812
The WS2812 devices (aka NeoPixel) are a popular component for art, design, and entertainment objects.

There are community libraries already:

- https://xod.io/libs/bradzilla84/neopixel/
- https://xod.io/libs/awgrover/adafruitneopixel/
- https://xod.io/libs/gabbapeople/christmas-tree/

But they use the full-size buffer internally (memory hogs) and not integrate with `xod/color`.

For casual use, a RAM-efficient node to set a single color on the whole strip is required. Also, a unified device object for further integration with bitmaps is necessary.

### xod-dev/ws2812/ws2812-device

Represents a chained sequence of WS2812 pixels.

| Dir | Pin  | Type    | Default | Description
| --- | ---- | ------- | ------- | -----------
| in  | PORT  | Port |     | The board port `DI` is connected to
| in  | W  | Number |  1   | Number of LEDs in the strip
| out | DEV | `self` |   | The device

### xod-dev/ws2812/fill-solid

Sets the color of all pixels in the device to the same value.

| Dir | Pin  | Type    | Default | Description
| --- | ---- | ------- | ------- | -----------
| in  | DEV  | ws2812-device |     | The device
| in  | C  | xod/color/color |  | Color value
| in | DO | Pulse | | Trigger to fill all pixels of the device
| out | DEV' | ws2812-device |   | The device
| out | DONE | Pulse | | Fires when the fill is done

### xod-dev/ws2812/clear

Clears (turns off) of all pixels in the device.

| Dir | Pin  | Type    | Default | Description
| --- | ---- | ------- | ------- | -----------
| in  | DEV  | ws2812-device |     | The device
| in | DO | Pulse | | Trigger to turn all pixels off
| out | DEV' | ws2812-device |   | The device
| out | DONE | Pulse | | Fires when the operation is done

### xod-dev/ws2812/ws2812-mono

A quickstart node to control a LED strip which is always filled with a single color

| Dir | Pin  | Type    | Default | Description
| --- | ---- | ------- | ------- | -----------
| in  | PORT  | Port |     | The board port `DI` is connected to
| in  | W  | Number |  1   | Number of LEDs in the strip
| in  | C  | xod/color/color |  | Color value
| in | ACT | Boolean |  True | When set to True the LEDs follow changes of `C` as they happen
| out | OK | Pulse | | Fires when an update operation is done

### PR DESCRIPTION
**Pull request by brusherru**
It closes #1921 

Also, it contains a little fix related to #1922 (PR #1926)

### CODE REVIEW

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+#ifndef _delay_us
+#define _delay_us(us) delayMicroseconds(us)
+#endif
+
+class WS2812
+{
+public:
+  WS2812(uint16_t pin, uint32_t length);
+  // TODO: Add destructor that sets pin to INPUT
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.h]: Nope. We should not

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#define NS_TO_CYCLES(n) ((n) / NS_PER_CYCLE)
+
+#ifndef _delay_us
+#define _delay_us(us) delayMicroseconds(us)
+#endif
+
+class WS2812 {
+public:
+    WS2812(uint16_t pin, uint32_t length);
+    // TODO: Add destructor that sets pin to INPUT
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.h]:
```suggestion
```

Not a good idea

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.h [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#define NS_PER_CYCLE (NS_PER_SEC / CYCLES_PER_SEC)
+#define NS_TO_CYCLES(n) ((n) / NS_PER_CYCLE)
+
+#ifndef _delay_us
+#define _delay_us(us) delayMicroseconds(us)
+#endif
+
+class WS2812 {
+public:
+    WS2812(uint16_t pin, uint32_t length);
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.h]:
```suggestion
    WS2812(uint8_t port, uint32_t length);
```

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    void show();
+    uint32_t getLength();
+
+private:
+    uint32_t _length;
+#ifdef __AVR__
+    volatile uint8_t* _port;
+#else
+    decltype(portOutputRegister(0)) _port;
+#endif
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.h]: The magic should be annotated with a comment

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+// Set the specified pin up as digital out
+void WS2812::setup(uint16_t pin) {
+    pinMode(pin, OUTPUT);
+    digitalWrite(pin, LOW);
+    _port = portOutputRegister(digitalPinToPort(pin));
+#if defined(ESP8266) || defined(ESP32)
+    _bit = pin;
+#else
+    _bit = getBitPosition(digitalPinToBitMask(pin));
+#endif
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp]: The magic should be annotated.

I suspect calling it `_bit` in the ESP case is plainly wrong and misguiding.

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+class WS2812 {
+public:
+    WS2812(uint16_t pin, uint32_t length);
+    // TODO: Add destructor that sets pin to INPUT
+    void fill(uint8_t r, uint8_t g, uint8_t b);
+    void sendPixel(uint8_t r, uint8_t g, uint8_t b);
+    void show();
+    uint32_t getLength();
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.h]:
```suggestion
    uint32_t getLength() const;
```

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    sendByte(r);
+    sendByte(b);
+}
+
+// Just wait long enough without sending any bots to cause the pixels to latch and display the last sent frame
+void WS2812::show() {
+    _delay_us((RES / 1000UL) + 1); // Round up since the delay must be _at_least_ this long (too short might not work, too long not a problem)
+}
+
+// Display a single color on the whole string
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp]:
```suggestion
// Display a single color on the whole strip
```

Find and replace everywhere as well. String is the “Hello”, the LEDs is a _strip_.

File: workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            [offCycles] "I"(NS_TO_CYCLES(_OFF) - 2)); \
+    }
+
+#define SEND_BITS(_ON, _OFF, _PORT) \
+    SEND_BIT(_ON, _OFF, _PORT, 0);  \
+    SEND_BIT(_ON, _OFF, _PORT, 1);  \
+    SEND_BIT(_ON, _OFF, _PORT, 2);  \
+    SEND_BIT(_ON, _OFF, _PORT, 3);  \
+    SEND_BIT(_ON, _OFF, _PORT, 4);  \
+    SEND_BIT(_ON, _OFF, _PORT, 5);
```
Comment (nkrkv) [inline workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp]: Why 6, not 8? :thinking:

Comment (brusherru) [reply workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp]: Oops. Overlooked that the PORTB register has only six accessible bits and forgot about other ones. Fixed it.

**Pull request merged**